tests/framework/extensions/codecoverage/codecoverage.go

```diff
@@ -0,0 +1,181 @@
+package codecoverage
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/rancher/norman/types"
+	apiv1 "github.com/rancher/rancher/pkg/apis/provisioning.cattle.io/v1"
+	"github.com/rancher/rancher/tests/framework/clients/rancher"
+	v1 "github.com/rancher/rancher/tests/framework/clients/rancher/v1"
+	"github.com/rancher/rancher/tests/framework/extensions/clusters"
+	"github.com/rancher/rancher/tests/framework/extensions/kubeconfig"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kwait "k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+)
+
+const (
+	cattleSystemNameSpace = "cattle-system"
+	localCluster          = "local"
+	rancherCoverFile      = "ranchercover.out"
+	agentCoverFile        = "agentcover.out"
+	outputDir             = "cover"
+)
+
+func checkServiceIsRunning(dynamicClient dynamic.Interface) error {
+	return kwait.Poll(500*time.Millisecond, 2*time.Minute, func() (done bool, err error) {
+		_, err = dynamicClient.Resource(corev1.SchemeGroupVersion.WithResource("pods")).Namespace(cattleSystemNameSpace).List(context.Background(), metav1.ListOptions{})
+		if k8sErrors.IsInternalError(err) || k8sErrors.IsServiceUnavailable(err) {
+			return false, nil
```

Comment:
So if you see I'm specifically checking an InternalError and ServiceUnavailable error, I don't want to return the error because then the poll ends reporting one of the errors above, when we don't want to do that. It is done this way because I am essentially pinging rancher until it's available after updating the rancher image.

Reply:
So this error catch' purpose only to get out of the poll if there are these errors, totally makes sense to me.

Reply:
So I return an error then it exits the poll before it's ready, what I can do instead is return an err on an else, just in case another error occurs.